### PR TITLE
[TRTLLM-7015] [feat] Enable `prompt_logprobs` in pytorch backend

### PR DESCRIPTION
**Pull request by venkywonka**
## Summary by CodeRabbit

- New Features
  - Unified logprobs support across backends, including prompt logprobs on PyTorch.
  - Automatic capture of required logits when requesting prompt logprobs.
  - Option to clear stored context logits to reduce memory after use.
  - Improved streaming behavior with consistent logprobs reporting.
- Bug Fixes
  - More accurate and consistent logprobs accumulation and length handling across backends.
- Tests
  - Expanded test coverage for logprobs across TensorRT and PyTorch backends, including streaming scenarios.

## Description

also fixes https://nvbugs/5516660

## Test Coverage

```yaml
- test_llm_pytorch.py::test_llm_return_logprobs
- test_llm_pytorch.py::test_llm_return_logprobs_streaming
```

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

### CODE REVIEW

File: docs/source/blogs/media/tech_blog10_baseline_performance_detail.png
Comment (venkywonka) [inline docs/source/blogs/media/tech_blog10_baseline_performance_detail.png:1]: these deletions were by mistake, will remove them from diff

File: tests/unittest/llmapi/test_llm.py [hunk 1]
```diff
@@ -1877,17 +1877,34 @@ async def main():
         asyncio.run(main())
 
 
-@force_ampere
+# @force_ampere
```
Comment (venkywonka) [inline tests/unittest/llmapi/test_llm.py:1880]: need to uncomment this

File: tensorrt_llm/sampling_params.py [hunk 2]
```diff
@@ -449,6 +450,16 @@ def _get_output_config(self, is_pytorch_backend: bool = False) -> tllme.OutputCo
 
         if is_pytorch_backend:
             config_kwargs["return_log_probs"] = bool(self.logprobs)
+            # For PyTorch backend: if prompt_logprobs is requested but return_context_logits is False,
+            # we need to internally enable context logits for prompt logprobs computation
+            # They will be dropped after computation if the user didn't explicitly request them
```
Comment (venkywonka) [inline tensorrt_llm/sampling_params.py:455]: comment unnecessary

File: tensorrt_llm/executor/result.py [hunk 3]
```diff
@@ -698,12 +704,12 @@ def compute_logprobs(
...
+    Compute top-K logprobs from logits when engine doesn't provide them directly.
 
-    Returns:
-        LogProbsResult, a NamedTuple containing:
-            - prompt: Optional[List[Dict[token_id, Logprob]]] logprobs for prompt tokens.
-            - generation: Optional[List[Dict[token_id, Logprob]]] logprobs for generated tokens.
+    Used for post-processing logits into logprobs.
+    - Prompt logprobs (from context_logits): always used.
+    - Generation logprobs (from generation_logits, TRT backend): used when backend doesn't compute them in sampler (e.g., TRT).
+    - Generation logprobs (PyTorch backend): not used; computed in sampler, not here.
```
Comment (venkywonka) [inline tensorrt_llm/executor/result.py:712]: include the `Returns` info

File: tests/unittest/llmapi/test_llm.py [hunk 4]
```diff
@@ -1877,21 +1877,36 @@ async def main():
...
+                             return_generation_logits: bool, backend: str):
+    llm_return_logprobs_test_harness(prompt_logprobs,
+                                     logprobs,
                                      return_context_logits,
-                                     return_generation_logits)
+                                     return_generation_logits,
+                                     backend=backend)
 
 
-@pytest.mark.skip(reason="https://nvbugs/5516660")
```
Comment (venkywonka) [inline tests/unittest/llmapi/test_llm.py:1894]: waiving this too as was able to fix it in the process of the PR. @chzblych  @hchings @LinPoly

File: tensorrt_llm/sampling_params.py [hunk 5]
```diff
@@ -449,6 +450,13 @@ def _get_output_config(self, is_pytorch_backend: bool = False) -> tllme.OutputCo
 
         if is_pytorch_backend:
             config_kwargs["return_log_probs"] = bool(self.logprobs)
+            if self.prompt_logprobs and not self.return_context_logits:
+                logger.info(
+                    "Since prompt_logprobs is requested but return_context_logits is False, "
+                    "internally enabling context logits for prompt logprobs computation. "
+                    "context logits will be dropped after computation as the user didn't explicitly request them."
+                )
+                config_kwargs["return_context_logits"] = True
```
Comment (dcampora) [inline tensorrt_llm/sampling_params.py:459]: I don't think this is very elegant. Can we pass the `prompt_logprobs` option into the framework and use the logic `prompt_logprobs or return_context_logits` where relevant instead?

**Review by dcampora — APPROVED** [review]
LGTM

**Review by LinPoly — APPROVED** [review]
LGTM, THX for fixing the bug.

File: tests/unittest/llmapi/test_llm.py [hunk 6]
```diff
@@ -1877,21 +1877,36 @@ async def main():
...
+        (2, None, False, False,
+         "trt"),  # prompt_logprobs without context_logits
+        (None, None, False, False, "trt"),  # no logprobs at all
+        # PyTorch backend test cases (adjusted for PyTorch limitations)
+        (2, None, True, False, "pytorch"
+         ),  # prompt_logprobs with context_logits
+        (None, 1, False, False,
+         "pytorch"),  # generation logprobs only (top-1, PyTorch limit)
+        (2, None, False, False,
+         "pytorch"),  # prompt_logprobs without context_logits
```
[the remark this answers is not in the document]
Comment (venkywonka) [reply tests/unittest/llmapi/test_llm.py:1896]: yep, i'll try that. my understanding is the `test_llm_pytorch.py` is for pytorch-only - but `test_llm.py` is for both

File: tensorrt_llm/sampling_params.py (hunk 5 above)
Comment (venkywonka) [reply tensorrt_llm/sampling_params.py:459]: by "into the framework" you mean touch the executor bindings (OutputConfig) and llm request data structures right? I actively avoided it due to the unnecessary plumbing complexity it would incur for this.

**Comment by venkywonka** [comment]
/bot run --extra-stage "A100X-TensorRT-Post-Merge-1, A100X-TensorRT-Post-Merge-2, A100X-TensorRT-Post-Merge-3, A100X-TensorRT-Post-Merge-4, A100X-TensorRT-Post-Merge-5, A100X-TensorRT-Post-Merge-6"

**Comment by venkywonka** [comment]
/bot run --extra-stage "A100X-TensorRT-Post-Merge-1, A100X-TensorRT-Post-Merge-2, A100X-TensorRT-Post-Merge-3, A100X-TensorRT-Post-Merge-4, A100X-TensorRT-Post-Merge-5, A100X-TensorRT-Post-Merge-6"

File: tests/unittest/llmapi/test_llm.py [hunk 7]
```diff
@@ -1804,95 +1804,110 @@ def llm_return_logprobs_test_harness(prompt_logprobs: Optional[int],
...
+                                     logprobs,
                                      return_context_logits,
-                                     return_generation_logits)
+                                     return_generation_logits,
+                                     backend=backend)
 
 
-@pytest.mark.skip(reason="https://nvbugs/5516660")
-@force_ampere
+# @force_ampere
```
Comment (venkywonka) [inline tests/unittest/llmapi/test_llm.py:1910]: uncomment this.

File: tensorrt_llm/sampling_params.py (hunk 5 above)
Comment (venkywonka) [reply tensorrt_llm/sampling_params.py:466]: added a TODO for this for now

File: tensorrt_llm/executor/result.py [hunk 8]
```diff
@@ -244,22 +244,32 @@ def _handle_sequence(self,
...
-            # handle logprobs directly from response tensors
+        elif response_tensors.log_probs is not None:  # PyTorch backend
+            # handle logprobs directly from response tensors given by sampler
             output._last_logprobs_len = len(output.logprobs)
-            output.logprobs = response_tensors.log_probs[src_idx]
+            # In streaming mode, since out-of-order responses are not possible,
+            # each streamed response_tensors.log_probs[src_idx]
+            # contains a streamwise monotonically growing list of logprobs.
+            # so we need to accumulate only the new ones unique to that particular streamed response
+            output.logprobs += response_tensors.log_probs[src_idx][
```
Comment (venkywonka) [inline tensorrt_llm/executor/result.py:262]: @hchings @LinPoly this lack of accumulation was causing https://nvbugs/5516660 , fixed here

File: tests/unittest/llmapi/test_llm.py [hunk 9]
```diff
@@ -1804,94 +1804,109 @@ def llm_return_logprobs_test_harness(prompt_logprobs: Optional[int],
...
         **llm_args_extra,
     )
 
-    prompts = ["A B C D E F G H I J K"]
-    sampling_params = SamplingParams(
-        logprobs=logprobs,
-        prompt_logprobs=prompt_logprobs,
-        return_context_logits=return_context_logits,
-        return_generation_logits=return_generation_logits)
+    with llm:
```
Comment (Superjomn) [inline tests/unittest/llmapi/test_llm.py:1827]: How about restoring this to use the LLM directly without the `with-`statement? The LLM instance should be able to clean up without requiring an explicit `__exit__` or `shutdown`. Introducing a `with-`statement might obscure issues that prevent proper destruction of the LLM instance.

**Comment by venkywonka** [comment]
@coderabbitai full review

File: tests/unittest/llmapi/test_llm.py (hunk 9 above)
Comment (venkywonka) [reply tests/unittest/llmapi/test_llm.py]: Ah thanks for the catch @Superjomn , i have undid my changes. i initially did it for a test that should fail, but that test is no longer needed.

**Review by achartier — APPROVED** [review]

**Review by Superjomn — APPROVED** [review]
LGTM
